lightning/src/chain/channelmonitor.rs

```diff
@@ -2106,6 +2113,74 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
+///
+/// Note that for every new monitor, you **must** persist the new ChannelMonitor
+/// to disk/backups. And, on every update, you **must** persist either the
+/// ChannelMonitorUpdate or the updated monitor itself. Otherwise, there is risk
+/// of situations such as revoking a transaction, then crashing before this
+/// revocation can be persisted, then unintentionally broadcasting a revoked
+/// transaction and losing money. This is a risk because previous channel states
+/// are toxic, so it's important that whatever channel state is persisted is
+/// kept up-to-date.
+///
+/// There are several different strategies for persistence, with different
+/// requirements. While not exhaustive, here are a few. You could:
+///   * have multiple redundant backups where you want to ensure that every
+///     backup is always accessible. In this case, you need to ensure that every
+///     copy is updated at each step before moving forward, with none falling
+///     out of sync,
+///   * have backups where you're OK with relying on only M of N backups
+///     temporarily, in which you don't care about consensus between backups and
+///     only needs to ensure that you read from the backup(s) with the highest
+///     [`ChannelMonitorUpdate::update_id`].
+///
+/// [`ChannelMonitorUpdate::update_id`]: struct.ChannelMonitorUpdate.html#structfield.update_id
+pub trait Persist<Keys: ChannelKeys>: Send + Sync {
+	/// Persist a new channel's data. The data can be stored with any file
+	/// name/path, but the identifier provided is the channel's outpoint. Note
+	/// that you **must** persist every new monitor to disk. See the Persist trait
+	/// documentation for more details, and for information on consensus between
+	/// backups.
+	///
+	/// See [`ChannelMonitor::write_for_disk`] for writing out a ChannelMonitor,
+	/// and [`ChannelMonitorUpdateErr`] for requirements when returning errors.
+	///
+	/// [`update_id`]: struct.ChannelMonitorUpdate.html#structfield.update_id
+	/// [`ChannelMonitor::write_for_disk`]: struct.ChannelMonitor.html#method.write_for_disk
+	/// [`ChannelMonitorUpdateErr`]: enum.ChannelMonitorUpdateErr.html
+	fn persist_new_channel(&self, id: OutPoint, data: &ChannelMonitor<Keys>) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Update one channel's data. The provided ChannelMonitor has already
+	/// applied the given update.
+	///
+	/// Note that on every update, you **must** persist either the
+	/// ChannelMonitorUpdate or the updated monitor itself to disk/backups. See
+	/// the Persist trait documentation for more details, and for information on
+	/// consensus between backups.
+	///
+	/// If an implementer chooses to persist the updates only, they need to make
+	/// sure that all the updates are applied to the ChannelMonitors *before* the
+	/// set of channel monitors is given to the ChainMonitor at startup time.
+	/// If full ChannelMonitors are persisted, then there is no need to persist
+	/// individual updates.
```

Comment:
Do we document somewhere that in case of recovery, you must restore offchain state before re-connecting the missed blocks ? Otherwise you may miss reaction to take due to not having the state.

AFAICT, saving either the whole ChannelMonitor or per-update is safe. If you go for the former, you should timestamps with the block height, otherwise you won't know from when to re-connect blocks ?

Reply:
ChannelMonitors save their latest block hash (and the ChannelManager docs describe what you mention here, though I agree it maybe should be somewhere else), so I don't think this can ever be an issue - eg if you just get the stream of updates you can replay from arbitrarily far back in the chain after applying the updates.

Reply:
> eg if you just get the stream of updates you can replay from arbitrarily far back in the chain after applying the updates (I've likely missed something here, though, feel free to correct :) ).

Right, a block timestamps is more a performance improvement rather than a security one. To avoid replay blocks for nothing.

Reply:
Right, in some module-level docs somewhere or something we should maybe mention you should consider serializing things out after connecting blocks, but its not required.